### PR TITLE
add static ip to kube play

### PR DESCRIPTION
**Pull request by mier85**
Signed-off-by: Michael Ernst <mier85@gmail.com>

this is implementing https://github.com/containers/podman/issues/8442

tried running lint & validate & tests, but they all failed before and after my commit. ( even tried the gate which is still partially in the CONTRIBUE.md but then realized that this seems to be no longer a thing )

### CODE REVIEW

**Comment by zhangguanzhang** [comment]
Lack of documentation and unit tests

**Comment by mier85** [comment]
@zhangguanzhang I will add the documentation in https://github.com/containers/podman/blob/master/docs/source/markdown/podman-play-kube.1.md and some unit tests in the file you pointed to.

**Comment by zhangguanzhang** [comment]
👌

File: cmd/podman/play/kube.go [hunk 1]
```diff
@@ -23,6 +24,7 @@ type playKubeOptionsWrapper struct {
 	TLSVerifyCLI   bool
 	CredentialsCLI string
 	StartCLI       bool
+	StaticIPCLI    string
```
Comment (Luap99) [inline cmd/podman/play/kube.go:27]: Not needed see other comment

File: cmd/podman/play/kube.go [hunk 2]
```diff
@@ -75,6 +77,10 @@ func init() {
 	flags.StringVar(&kubeOptions.Authfile, authfileFlagName, auth.GetDefaultAuthFile(), "Path of the authentication file. Use REGISTRY_AUTH_FILE environment variable to override")
 	_ = kubeCmd.RegisterFlagCompletionFunc(authfileFlagName, completion.AutocompleteDefault)
 
+	staticIPFlagName := "ip"
+	flags.StringVar(&kubeOptions.StaticIPCLI, staticIPFlagName, "", "Static IP address to assign to this pod")
+	_ = kubeCmd.RegisterFlagCompletionFunc(staticIPFlagName, completion.AutocompleteDefault)
+
```
Comment (Luap99) [inline cmd/podman/play/kube.go:83]: You can just use an IP flag. No need for the extra parsing logic
```suggestion
	staticIPFlagName := "ip"
	flags.IPVar(&kubeOptions.StaticIP, staticIPFlagName, nil, "Static IP address to assign to this pod")
	_ = kubeCmd.RegisterFlagCompletionFunc(staticIPFlagName, completion.AutocompleteNone)

```

File: cmd/podman/play/kube.go [hunk 3]
```diff
@@ -119,6 +125,14 @@ func kube(cmd *cobra.Command, args []string) error {
...
 	}
 
+	if kubeOptions.StaticIPCLI != "" {
+		ipAddress := net.ParseIP(kubeOptions.StaticIPCLI)
+		if ipAddress == nil {
+			return fmt.Errorf("failed parsing ip address: %s", kubeOptions.StaticIPCLI)
+		}
+		kubeOptions.StaticIP = ipAddress
+	}
+
```
Comment (Luap99) [inline cmd/podman/play/kube.go:135]: Not needed see other comment

**Review by Luap99 — COMMENTED** [review]
Thanks for the contribution @mier85 

I left some comment to simplify this a bit.

`podman play kube` can be run via the api and podman-remote. So you also need to add this option to the api and remote client.

The api parameter is defined here: https://github.com/containers/podman/blob/0c2a43b99db8a4fd75412a277da6de2731017d3e/pkg/api/handlers/libpod/play.go

Documentation for the api parameter has to be added here: https://github.com/containers/podman/blob/0c2a43b99db8a4fd75412a277da6de2731017d3e/pkg/api/server/register_play.go

Parameter for the remote client has to be added here: https://github.com/containers/podman/blob/master/pkg/bindings/play/play.go

File: cmd/podman/play/kube.go (hunk 1 above)
Comment (mier85) [reply cmd/podman/play/kube.go:27]: flags will make things much easier, will update that

**Comment by rhatdan** [comment]
@mier85 Still working on this?  We need to get this in by next week, for the 3.0 release, if we are going to get it done.

**Comment by mier85** [comment]
@rhatdan still working on it. Will try to get this done ASAP this week. 

But I have a question. As several Pods could be spawned when deploying ( e.g. replicas in a kubernetes manifest ) , a single static IP wouldn't work. In this case, should be assign a range starting with the IP or should we fail and ask for several IPs or should this be a "single container only" option?

**Comment by rhatdan** [comment]
I am no kubernetes expert, but I would figure fail would be the correct thing to do.

**Comment by mheon** [comment]
Making static IPs incompatible with Deployments, at least initially, sounds like a fine solution.

**Comment by hoonetorg** [comment]
@mier85: i see it that way: 
if kubernetes yaml is of type pod, only one pod will be deployed. if the kind is deployment than replicas is one, also only 1 container can be deployed. that are two valid scenarios for play kube, where a static ip could be used.
the thing with the range is probably a bit too difficult for the initial shot.

**Comment by vrothberg** [comment]
Friendly ping. Are we good to continue?

**Comment by rhatdan** [comment]
@umohnani8 Could you take this one over?

**Comment by Luap99** [comment]
Replaced by #10043

**Pull request closed without merging**